### ISSUE TITLE
SWRL rules with incorrect equals()

### ISSUE THREAD

**Issue by Galigator** [owlcs/owlapi#512]
Hello,

While converting pellet to owl-api5 (5.0.2-SNAPSHOT git:7561524 )  I have hit the message of the title.

The test case is :
https://github.com/Galigator/pellet/blob/owlapi5/test/src/test/java/com/clarkparsia/explanation/test/AbstractExplanationTest.java
Line : 368

The error doesn't show when using owlapi4

OWL.classAssertion(a, A), 
OWL.propertyAssertion(a, dp, OWL.constant(9)),/
SWRL.rule(
	SWRL.antecedent(
	        SWRL.classAtom(A, x),
		SWRL.propertyAtom(dp, x, dx),
		SWRL.greaterThan(dx, SWRL.constant(5)),
		SWRL.lessThan(dx, SWRL.constant(10))
	),
	SWRL.consequent(SWRL.classAtom(B, x))
) 

Must entail OWL.classAssertion(a, B)

But the following stack trace show :

org.semanticweb.owlapi.model.OWLRuntimeException: Explanation contains removed axiom: DLSafeRule( Body(ClassAtom(<http://www.example.org/test#A> Variable(<http://www.example.org/test#x>)) DataPropertyAtom(<http://www.example.org/test#dp> Variable(<http://www.example.org/test#x>) Variable(<http://www.example.org/test#dx>)) BuiltInAtom(<http://www.w3.org/2003/11/swrlb#lessThan> Variable(<http://www.example.org/test#dx>) "10"^^xsd:integer ) BuiltInAtom(<http://www.w3.org/2003/11/swrlb#greaterThan> Variable(<http://www.example.org/test#dx>) "5"^^xsd:integer )) Head(ClassAtom(<http://www.example.org/test#B> Variable(<http://www.example.org/test#x>))) )
at com.clarkparsia.owlapi.explanation.HSTExplanationGenerator.recurse(HSTExplanationGenerator.java:300)
at com.clarkparsia.owlapi.explanation.HSTExplanationGenerator.constructHittingSetTree(HSTExplanationGenerator.java:242)
at com.clarkparsia.owlapi.explanation.HSTExplanationGenerator.getExplanations(HSTExplanationGenerator.java:120)
at com.clarkparsia.explanation.test.OWLAPIExplanationTest.testExplanations(OWLAPIExplanationTest.java:243)
at com.clarkparsia.explanation.test.AbstractExplanationTest.testExplanations(AbstractExplanationTest.java:132)
at com.clarkparsia.explanation.test.AbstractExplanationTest.testExplanations(AbstractExplanationTest.java:120)
at com.clarkparsia.explanation.test.AbstractExplanationTest.ruleBuiltinTest2(AbstractExplanationTest.java:385)

Many thanks for owl-api ! Lot of fun with it.

**Comment by ignazio1977** [comment]
HSTExplanationGenerator has not undergone any functional change between 4 and 5, so the change in behaviour must originate somewhere else.

A couple of possibilities: SWRL rules have head and body as lists, which are supposed to be ordered by insertion order - but this has had a number of changes between versions. The equals() method needs to take this into account, and might have failed to consider that two rules are equivalent even if the order differs.
Another issue was with object duplication, which did not duplicate annotations for SWRL rules. Some OWLAPI 4 versions would create SWRL rules that are identical except for annotations, and that would not trigger this exception. OWLAPI 5 and recent OWLAPI 4 might show this instead.

It's hard for me to replicate because your Pellet fork has dependencies on other non released forks, e.g., org.jgrapht:jgrapht-ext:jar:0.9.2-galigator

**Comment by Galigator** [comment]
Thx for your time. I have done more investigation on this. It looks like it is a problem into the GlassBoxExplanation (pellet) class, something probably related to the ordering of in body of rules.

**Comment by Galigator** [comment]
I have change the interface SWRLRule to sort the body/head from the components stream because it is use in the main comparison function. This solve the primary problems. Is sorting an option here ? "OWLAPI :: Tests and Tutorials" run well, but sorting can be time consuming.

	@Override
	default Stream<?> components()
	{
		return Stream.of(body().sorted(), head().sorted(), annotations());
	}

Solving this problem show annotations problems that come next :-)

**Comment by ignazio1977** [comment]
It's sorted by insertion order because of editor requirements: Users want to have the order of items be preserved.

We won't be able to change that, but we can change the equals() to be order agnostic - that should fix your issue. Otherwise the sorting should be done in the callers, although I see how that could be difficult.

All other collections/streams inside OWLObjects are created from sorted lists, so their order is predictable.

**Comment by ignazio1977** [comment]
I've gone for a slightly different version - rather than changing the interface, I've changed the implementation only, as that's the same place where the ordering of head and body is enforced (theoretically another implementation of SWRLRule could choose to sort head and body differently, they're defined to be sets with no particular order).

**Comment by Galigator** [comment]
Thx for this fast change.

Every things is fine now.

**Comment by melaasar** [comment]
I ran into a similar (or the same) exception in v5.1.16:
https://github.com/owlcs/owlapi/blob/d669960f5d75527c00b391e2997df6f06ec5122a/tools/src/main/java/com/clarkparsia/owlapi/explanation/HSTExplanationGenerator.java#L294-L295

I noticed the comment above it saying "// How can this be the case???". Does this mean it is not an expected situation? What is the situation that would cause such exception to be thrown? and how can it be fixed? Thanks a lot!
